src/python/pants/engine/rules.py

```diff
@@ -420,49 +469,70 @@ def dependency_optionables(self):
     return tuple()
 
 
-class RuleIndex(datatype(['rules', 'roots'])):
+class RuleIndex(datatype(['rules', 'roots', 'union_rules'])):
   """Holds a normalized index of Rules used to instantiate Nodes."""
 
   @classmethod
-  def create(cls, rule_entries):
+  def create(cls, rule_entries, union_rules=None):
     """Creates a RuleIndex with tasks indexed by their output type."""
     serializable_rules = OrderedDict()
     serializable_roots = OrderedSet()
+    union_rules = OrderedDict(union_rules or ())
 
     def add_task(product_type, rule):
       if product_type not in serializable_rules:
         serializable_rules[product_type] = OrderedSet()
       serializable_rules[product_type].add(rule)
 
+    def add_root_rule(root_rule):
+      serializable_roots.add(root_rule)
+
     def add_rule(rule):
       if isinstance(rule, RootRule):
-        serializable_roots.add(rule)
-        return
-      # TODO: Ensure that interior types work by indexing on the list of types in
-      # the constraint. This heterogenity has some confusing implications:
-      #   see https://github.com/pantsbuild/pants/issues/4005
-      for kind in rule.output_constraint.types:
-        add_task(kind, rule)
-      add_task(rule.output_constraint, rule)
+        add_root_rule(rule)
+      else:
+        # TODO: Ensure that interior types work by indexing on the list of types in
+        # the constraint. This heterogenity has some confusing implications:
+        #   see https://github.com/pantsbuild/pants/issues/4005
+        for kind in rule.output_constraint.types:
+          add_task(kind, rule)
+        add_task(rule.output_constraint, rule)
+
+    def add_type_transition_rule(union_rule):
+      """
+      NB: This does not require that union bases be supplied to `def rules():`! not sure if that's
```

Comment:
Totally fine with that, for the same reason you put in the description.

Reply:
Have moved this to a comment noting that this is because the union type is never instantiated!